linera-core/src/client.rs

```diff
@@ -747,119 +731,102 @@ where
-        ensure!(
-            certificate.value().block() == Some(&content.block),
-            ChainClientError::BlockProposalError(
-                "A different operation was executed in parallel (consider retrying the operation)"
-            )
-        );
-        if content.round.is_fast() {
-            ensure!(
-                certificate.value().is_confirmed(),
-                ChainClientError::InternalError("Certificate is not for a confirmed block")
-            );
+            .communicate_chain_action(committee, submit_action, value)
+            .await?;
+        if certificate.value().is_confirmed() {
             self.process_certificate(certificate.clone(), vec![])
                 .await?;
             Ok(certificate)
         } else {
-            ensure!(
-                certificate.value().is_validated(),
-                ChainClientError::InternalError("Certificate is not for a validated block")
-            );
             self.finalize_block(committee, certificate).await
         }
     }
 
-    /// Broadcasts certified blocks and optionally one more block proposal.
-    /// The corresponding block heights should be consecutive and increasing.
+    /// Broadcasts certified blocks to validators.
     async fn communicate_chain_updates(
         &mut self,
         committee: &Committee,
         chain_id: ChainId,
-        action: CommunicateAction,
-    ) -> Result<Option<Certificate>, ChainClientError> {
+        height: BlockHeight,
+        delivery: CrossChainMessageDelivery,
+    ) -> Result<(), ChainClientError> {
         let storage_client = self.storage_client().await;
         let nodes: Vec<_> = self.validator_node_provider.make_nodes(committee)?;
-        let (maybe_hash_round, votes) = communicate_with_quorum(
+        communicate_with_quorum(
             &nodes,
             committee,
-            |value: &Option<LiteVote>| -> Option<_> {
-                value
-                    .as_ref()
-                    .map(|vote| (vote.value.value_hash, vote.round))
+            |_: &()| (),
+            |name, node| {
+                let mut updater = ValidatorUpdater {
+                    name,
+                    node,
+                    storage: storage_client.clone(),
+                };
+                Box::pin(async move {
+                    updater
+                        .send_chain_information(chain_id, height, delivery)
+                        .await
+                })
             },
+        )
+        .await?;
+        Ok(())
+    }
+
+    /// Broadcasts certified blocks and optionally a block proposal, certificate or
+    /// leader timeout request.
+    ///
+    /// In that case, it verifies that the validator votes are for the provided value,
+    /// and returns a certificate.
+    async fn communicate_chain_action(
+        &mut self,
+        committee: &Committee,
+        action: CommunicateAction,
+        value: HashedValue,
```

Comment:
I would still keep the value as part of the action for the one variant where it is needed.